marconi-mamba/examples/json-rpc-server/src/Main.hs

```diff
@@ -1,29 +1,98 @@
+    , "addr_test1qqwh05w69g95065zlr4fef4yfpjkv8r3dr9h3pu0egy5n7pwhxp4f95svdjr9dmtqumqcs6v49s6pe7ap4h2nv8rcaasgrkndk"
+    , "addr_test1qrr6urmwy2nle3xppnjg9xcukasrwfyfjv9eqh97km84ra53q4hau90tjeldx0mv9eka2z73t9727xl8jny3cy8zetqsdjctpd"
+    , "addr_test1wrn2wfykuhswv4km08w0zcl5apmnqha0j24fa287vueknasq6t4hc"
+    ,"addr_test1qr30nkfx28r452r3006kytnpvn39zv7c2m5uqt4zrg35mly35pesdyk43wnxk3edkkw74ak56n4zh67reqjhcfp3mm7qtyekt4"
+    , "addr_test1wr9gquc23wc7h8k4chyaad268mjft7t0c08wqertwms70sc0fvx8w"
+    , "addr_test1vqeux7xwusdju9dvsj8h7mca9aup2k439kfmwy773xxc2hcu7zy99"
+    ]
+defaultDbpath :: FilePath
+defaultDbpath = "./.marconidb/4/utxo-db"
 
--- TODO
-bech32Addresses :: String   -- ^  valid address to keep track of
-bech32Addresses = "bc1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3qccfmv3"
+data CliOptions = CliOptions
+    { _utxoPath  :: FilePath -- ^ path to utxo sqlite database
+    , _addresses :: TargetAddresses
+    }
 
--- TODO
-dbpath :: FilePath -- ^ valid SQLite marconi UTxo database path
-dbpath = "./utxoDB"
+cliParser :: Parser CliOptions
+cliParser = CliOptions
+    <$> strOption (long "utxo-db"
+                              <> short 'd'
+                              <> metavar "FILENAME"
+                              <>  showDefault
+                              <> value defaultDbpath
+                              <> help "Path to the utxo database.")
+    <*> pAddressesParser
 
-addresses :: TargetAddresses
-addresses = targetAddressParser bech32Addresses
+pAddressesParser :: Parser TargetAddresses
+pAddressesParser = targetAddressParser <$> strOption
+    (long "addresses-to-index"
+     <> metavar "Address"
+     <>  showDefault
+     <> value bech32DefaultAddresses
+     <> help ("White space separated list of addresses to index."
+                  <>  " i.e \"address-1 address-2 address-3 ...\"" ) )
 
 main :: IO ()
 main = do
-    putStrLn "Starting the Example rpc http-server on port 3000 example"
-    env <- bootstrapJsonRpc dbpath Nothing addresses
-    bootstrapHttp env
+    (CliOptions dbpath addresses) <- execParser $ info (cliParser <**> helper) mempty
+    putStrLn $ "Starting the Example RPC http-server:"
+        <>"\nport =" <> show (3000 :: Int)
+        <> "\nutxo-db =" <> dbpath
+        <> "\nnumber of addresses to index = " <> show (length addresses)
+
+    env <- bootstrapJsonRpc dbpath Nothing addresses Mainnet
+    race_ (bootstrapHttp env) (mocUtxoIndexer (env ^. queryEnv) )
+
+mocUtxoIndexer :: DBQueryEnv -> IO ()
+mocUtxoIndexer env = open "" (Depth 4) >>= innerLoop
+    where
+        utxoIndexer = env ^. queryComm . indexer
+        qreq = env ^. queryComm . queryReq
+        innerLoop ix = do
+            isquery <-  atomically $ tryReadTMVar qreq
+            unless (null isquery) $ bracket_
+                (atomically (takeTMVar qreq ) )
+                (atomically (takeTMVar qreq ) )
+                (atomically  (putTMVar utxoIndexer ix) )
```

Comment:
The synchronisation here looks expensive: when a query arrives, we wait for it to release a semaphore, to release an `ix` to release its action before we wait for the end of the transaction.
The first step looks unnecessary to me, but as always concurrency is such a mess that I may be wrong (if I'm right removing the initialisation in this bracket and the one in the query shouldn't impact the behaviour and save us a bit of time).

Reply:
Thanks for taking a detail look at this part, as I could certainly use another pair of eyes here.
The short answer is you're correct and the initialization is unnecessary. I will remove it as it is distracting

The longer answer is section of the code is mocking the Marconi [indexer.hs](https://github.com/input-output-hk/plutus-apps/blob/1009/report-marconi-target-addresses-v2/marconi/src/Marconi/Indexers.hs#L180) insert  process, so the code resembles that logic.